### ISSUE TITLE
Pie in the sky: a new build system

### ISSUE THREAD

**Issue by nicowilliams** [heimdal/heimdal#693]
Why not.  It'd be nice to have a build system that handles Unix and Windows (native, cygwin, whatever).  It'd be nice if the Windows build could be made to use clang.  All without having to keep Makefiles and NTMakefiles or .sln and .vcxproj files in sync.

Pie in the sky.

However, there is [Ninja](https://ninja-build.org/) and the [higher-level tools built on Ninja](https://github.com/ninja-build/ninja/wiki/List-of-generators-producing-ninja-build-files), of which several claim to support multi-platform, Unix/Linux/Windows, of which [Meson](https://mesonbuild.com/) looks promising, though so do others.  The plethora of systems makes it hard to know where to start.

One thing that would be nice is if we could ditch the recursive Makefile pattern in moving to Ninja (or anything else).  Another would be moving to clang on Windows, so we can finally make full use of C99 and even C11.  Yet another nice thing would be leaving libtool behind.

Well, this will no doubt take a lot of energy.  There are a lot of complications...  For one, how does autoconf mix with whatever new build system?  My preference would be to have autoconf generate configurations  that we can commit in-tree, and the to build one would select a configuration or else make a new one, and then kick off a build, which in turn would speed up Travis-CI builds.  Another complication is testing.  Another is code coverage measurement.  There's a lot of complications that make it easier to stay with what we have even though it's got some serious downsides (two different build systems to keep in sync, recursive make, no easy way to move to clang on Windows...).

**Comment by abartlet** [comment]
I've not used it, but I've been to talks about Meason (even got a mug
for asking good questions!) at linux.conf.au.  It seems to have gained
some traction.

Samba has yet another build system for Heimdal, being waf.  It doesn't
built all of it, just enough to integrate our vendored copy.

Some lessons from that:
 * cross-compilation is harder than it looks
 * Heimdal is a bit of a special case (compared with many others) with
the need to compile the ASN.1 compiler first, and with the host (not
target) CC.
 * A Python based build system has worked well for Samba.  While our
build system (waf) is far, far more than 'just a python script', being
able to fall back to python syntax for stuff is very helpful
 * Life in parallel is a pain, get in and get out, don't maintain both
for ages.

The transition is hard, here are a few things that might help:
 * Compare CPP build outputs with tools like my gcc-E script: 
https://wiki.samba.org/index.php/Testing_removal_of_ifdef_or_configure_checks#A_script_to_help_obtain_CPP_output
 * Try to make the config.h the same in both tools, for a number of
platforms
 * you may wish to remove unused checks first.  Samba still has many
un-used checks from code long removed
 * Use some of the library checking tools to check you still produce
the same ABI and symbol versions etc. 

Finally, if using a python tool:
 * Go pure python3
 * Samba provides a transitional shell ./configure and make-based
Makefile.  This is still how most people build samba, but the wrapper
is trivial, take a look.

**Comment by lhoward** [comment]
My two cents, it feels like a rabbit warren, we'd be better focusing on other things. But I'm happy to adopt a new build system if someone else doe the work ;)
